### ISSUE TITLE
LUKS: crashes with synthetic inputs with ASan

### ISSUE THREAD

**Issue by AlekseyCherepanov** [openwall/john#5070]
While fuzzer gets killed by OOM (#4979) it is possible to crash manually:

`keybytes` 0 and `stripes` 0 taken from binary header:
```
$ python -c 'print "$luks$1$592$" + "4c554b53babe00" + "0" * 1170 + "$0$A$" + "b" * 40' > t.pw
$ ./run/john --format=luks t.pw
[...]
==21413==ERROR: AddressSanitizer: FPE on unknown address 0x55b1a44ea869 (pc 0x55b1a44ea869 bp 0x7ffcbd237d60 sp 0x7ffcbd237d20 T0)
    #0 0x55b1a44ea868 in AF_merge /home/user/john/src/luks_fmt_plug.c:184
    #1 0x55b1a44ec3fe in crypt_all /home/user/john/src/luks_fmt_plug.c:582
```

Insufficient data:
```
$ python -c 'print "$luks$1$592$" + "4c554b53babe000161657300000000000000000000000000000000000000000000000000000000006362632d65737369763a73686132353600000000000000000000000000000000736861310000000000000000000000000000000000000000000000000000000000000408000000104f386b50df3fcd9132589a934851faaff16709ff628ed0b628b0d7151b3600c0b3f95d8404a8b35fdf5dd6b6ff10f4c352fde11900010b1762663664393836622d633836352d343261622d616534662d6165313336633938383735360000000000ac71f3000430f5d9c39e349b48d7cf1771d9c152840b389a4353ff186436ec75cc397529ed40260000000800000fa00000dead0000000000000000000000000000000000000000000000000000000000000000000000000000008800000fa00000dead0000000000000000000000000000000000000000000000000000000000000000000000000000010800000fa00000dead0000000000000000000000000000000000000000000000000000000000000000000000000000018800000fa00000dead0000000000000000000000000000000000000000000000000000000000000000000000000000020800000fa00000dead0000000000000000000000000000000000000000000000000000000000000000000000000000028800000fa00000dead0000000000000000000000000000000000000000000000000000000000000000000000000000030800000fa00000dead0000000000000000000000000000000000000000000000000000000000000000000000000000038800000fa0" + "$64000$A$" + "b" * 40' > t.pw
$ ./run/john --format=luks t.pw
[...]
==21937==ERROR: AddressSanitizer: SEGV on unknown address 0x625000010000 (pc 0x7f696d7f2c3d bp 0x000000000000 sp 0x7ffe9b2daab8 T0)
==21937==The signal is caused by a READ memory access.
    #0 0x7f696d7f2c3c in AES_decrypt (/lib/x86_64-linux-gnu/libcrypto.so.1.1+0x8fc3c)
    #1 0x7f696d8f3815 in CRYPTO_cbc128_decrypt (/lib/x86_64-linux-gnu/libcrypto.so.1.1+0x190815)
    #2 0x55b25fab01d5 in decrypt_aes_cbc_essiv /home/user/john/src/luks_fmt_plug.c:241
```

**Comment by AlekseyCherepanov** [comment]
BTW result of `fread` is checked incorrectly. Its return type is `size_t`. `readbytes` is a misleading name. Actually its value is either 0 or 1 due to order of arguments.
```c
	readbytes = fread(*cp, afsize, 1, myfile);

	if (readbytes < 0) {
```

`LUKS` is another format that can read files by names specified in ciphertexts (discussion for TrueCrypt: #5068). But the file is opened in `valid` and does not cause `exit`.

```
$ python -c 'print "$luks$0$592$" + "4c554b53babe00" + "0" * 1170 + "$0$B$A$" + "b" * 40' > t.pw
$ ./run/john --format=luks t.pw
A : No such file or directory!
[...]
Direct leak of 1244 byte(s) in 1 object(s) allocated from:
    #0 0x7f2a02dbd810 in __interceptor_strdup ../../../../src/libsanitizer/asan/asan_interceptors.cc:405
    #1 0x55d15df188f8 in xstrdup /home/user/john/src/memory.c:136
    #2 0x55d15dd8c7f6 in valid /home/user/john/src/luks_fmt_plug.c:346
```

Leak is caused by `return` instead of `goto err`:
```c
		if (hash_plugin_parse_hash(q, &buf, afsize, 0) == -1) {
			return 0;
		}
```

**Comment by AlekseyCherepanov** [comment]
It looks like call to `hash_plugin_parse_hash` with `afsize` in `valid` is wrong. Other field is used in `get_salt`.

**Comment by AlekseyCherepanov** [comment]
`afsize` is passed to `hash_plugin_parse_hash` as `int`:
```c
	// can this go over 4gb?
	*cp =(unsigned char*) mem_calloc(1, afsize + 1);
	if (!*cp)
		goto bad;
```

Sample to trigger wrapping over `INT_MAX`:
```
$ touch A
$ python -c 'print "$luks$0$592$" + "4c554b53babe" + "0000" + "0" * (32 * 2 * 3) + "00000000" + "7fffffff" + "00000001" * 120 + "$2147483647$B$A$" + "b" * 40' > t.pw && ./run/john --format=luks t.pw
[...]
==3379==WARNING: AddressSanitizer failed to allocate 0xffffffff80000000 bytes
==3379==AddressSanitizer's allocator is terminating the process instead of returning 0
```

Without ASan it would exit safely in `mem_calloc`.

**Comment by AlekseyCherepanov** [comment]
It seem in `valid` there is attempt to check that `keybytes * stripes` fits into `uint32_t`:
```c
#define john_ntohl(x) ((((x)>>24) & 0xffL) | (((x)>>8) & 0xff00L) | \
		(((x)<<8) & 0xff0000L) | (((x)<<24) & 0xff000000L))
[...]
struct luks_phdr {
[...]
	uint32_t keyBytes;
[...]
		uint32_t stripes;
[...]
};
[...]
	if ( (uint64_t)(john_ntohl(cs.myphdr.keyBytes)*john_ntohl(cs.myphdr.keyblock[bestslot].stripes)) !=
		keybytes*stripes)
		goto err;
```

This check does not work. I guess `john_ntohl(...) * john_ntohl(...)` was intended to work in `uint32_t`, but `0xffL` in the macro upgrades type to `long`. So there is undefined behaviour because product of two `uint32_t` can overflow signed `long`.

But then there is a field that should be equal to the product and is checked with `isdec` that limits at `INT_MAX`.
```c
	int res;
[...]
	uint64_t keybytes, stripes;
[...]
	if (!isdec(p))
		goto err;
	res = atoi(p);
	if (res != keybytes*stripes)
		goto err;
```

**Comment by AlekseyCherepanov** [comment]
`1 * 1` with 0 bytes of data:
```
$ python -c 'print "$luks$1$592$" + "4c554b53babe" + "0000" + "0" * (32 * 2 * 3) + "00000000" + "00000001" + "00000001" * 120 + "$1$A$" + "b" * 40' > t.pw && ./run/john --format=luks t.pw
[...]
==14438==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffd360825f0 at pc 0x5641fb12ed39 bp 0x7ffd36081c40 sp 0x7ffd36081c38
WRITE of size 1 at 0x7ffd360825f0 thread T0
    #0 0x5641fb12ed38 in XORblock /home/user/john/src/luks_fmt_plug.c:150
    #1 0x5641fb12fa1d in AF_merge /home/user/john/src/luks_fmt_plug.c:194
    #2 0x5641fb1313fe in crypt_all /home/user/john/src/luks_fmt_plug.c:583
```

**Comment by AlekseyCherepanov** [comment]
`af_sectors` rounds `afsize` up to next multiple of 512. So non-multiple should not be allowed.

**Comment by AlekseyCherepanov** [comment]
It looks like `luks2john.py` can populate incomplete sectors:
```python
    afsize = af_sectors(keyBytes, stripes)
[...]
    cipherbuf = f.read(afsize)
```

`f.read()` can return less bytes than `afsize`. Result is not checked.

It does not seem practical to fix all code to support incomplete sectors.

Self-tests suggest that it is not expected case. So I guess rejection in `valid` without warning would be ok.

**Comment by AlekseyCherepanov** [comment]
Crash with 512 bytes of data provided accurately (lines numbers are wrong due to my changes):
```
$ python -c 'print "$luks$1$592$" + "4c554b53babe" + "0000" + "0" * (32 * 2 * 3) + "00000000" + "00000200" + "00000001" * 120 + "$512$" + "A" * 683 + "=" + "$" + "b" * 40' > t.pw && ./run/john --format=luks t.pw
[...]
==26871==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffe7e691b00 at pc 0x55e6ee56bfaf bp 0x7ffe7e690e40 sp 0x7ffe7e690e38
WRITE of size 1 at 0x7ffe7e691b00 thread T0
    #0 0x55e6ee56bfae in pbkdf2_sha1_sse /home/user/john/src/pbkdf2_hmac_sha1.h:301
    #1 0x55e6ee56c3d3 in crypt_all /home/user/john/src/luks_fmt_plug.c:595
```

The formula for `size` seems wrong:
```c
		size = strlen(p) / 4 * 3 + 1;
		buf = mem_calloc(1, size+4);
		base64_convert(p, e_b64_mime, strlen(p), buf, e_b64_raw, size+4, flg_Base64_NO_FLAGS, 0);

		cs.afsize = size;
```

```
>>> len(('a' * 512).encode('base64').replace('\n', ''))
684
>>> 684 / 4 * 3 + 1
514
```

**Comment by AlekseyCherepanov** [comment]
I tried to fix the size. It did not not affect the crash.

Actual problem is that `dklen` should be '<= sizeof(keycandidate[0])':
```c
		int dklen = john_ntohl(cur_salt->myphdr.keyBytes);
		uint32_t keycandidate[MIN_KEYS_PER_CRYPT][256/4];
```

**Comment by solardiz** [comment]
> ```c
> #define john_ntohl(x) ((((x)>>24) & 0xffL) | (((x)>>8) & 0xff00L) | \
> 		(((x)<<8) & 0xff0000L) | (((x)<<24) & 0xff000000L))
> ```

I think regardless of your other findings the above should be fixed to be either `JOHN_SWAP(x)` or `(x)` depending on machine endianness. Ditto for similar macros in `dmg2john.c`. In fact, the macros could be added to `johnswap.h` (under `john_` prefixed names) and that file included from both `.c` files.

Similarly, the `ll` form only present/used in `dmg2john.c` should use `JOHN_SWAP64`.

**Comment by solardiz** [comment]
For any changes to `dmg2john.c`, please test them against files in our `john-samples` repo.

**Comment by AlekseyCherepanov** [comment]
`valid`:
```c
	int is_inlined, i, bestslot=0;
[...]
	unsigned int bestiter = 0xFFFFFFFF;
[...]
	for (i = 0; i < LUKS_NUMKEYS; i++) {
			if ((john_ntohl(cs.myphdr.keyblock[i].passwordIterations) < bestiter)
			&& (john_ntohl(cs.myphdr.keyblock[i].passwordIterations) > 1) &&
			(john_ntohl(cs.myphdr.keyblock[i].active) == 0x00ac71f3)) {
				bestslot = i;
				bestiter =
				john_ntohl(cs.myphdr.keyblock[i].passwordIterations);
			}
	}
```

`get_salt`:
```c
	struct custom_salt_LUKS cs, *psalt;
[...]
	unsigned int bestiter = 0xFFFFFFFF;
[...]
	memset(&cs, 0, sizeof(cs));
[...]
	for (cnt = 0; cnt < LUKS_NUMKEYS; cnt++) {
			if ((john_ntohl(cs.myphdr.keyblock[cnt].passwordIterations) < bestiter)
			&& (john_ntohl(cs.myphdr.keyblock[cnt].passwordIterations) > 1) &&
			(john_ntohl(cs.myphdr.keyblock[cnt].active) == 0x00ac71f3)) {
				cs.bestslot = cnt;
				cs.bestiter =
				john_ntohl(cs.myphdr.keyblock[cnt].passwordIterations);
			}
	}
```

There are multiple keys and `john` chooses one that is active and has least number of iterations. So if there is not active key, then incorrect value would be used. So I guess it should be rejected.

Variable `bestiter` is not updated in `get_salt`. So it works as intended when best key is the last (e.g. there is only 1 active key).

**Comment by AlekseyCherepanov** [comment]
Hm, `bestslot` affects value in `stripes`, so the difference in chosen key between `valid` and `get_salt` is a potential for buffer overflow.

**Comment by AlekseyCherepanov** [comment]
`passwordIterations` in binary structure has type `uint32_t`. Type of variables `bestiter` is `unsigned int`. Field `bestiter` in `custom_salt_LUKS` is `int`. It frustrates...

**Comment by AlekseyCherepanov** [comment]
Oh, comments mention possible redesign (#557).

**Comment by AlekseyCherepanov** [comment]
`ntohll` in `dmg2john.c` is defined outside of endianness check, so I guess `dmg2john.c` should work incorrectly in BE environment.

**Comment by solardiz** [comment]
> `ntohll` in `dmg2john.c` is defined outside of endianness check, so I guess `dmg2john.c` should work incorrectly in BE environment.

I guess so. Please fix that after we merge #5073.
